### ISSUE TITLE
The numerial difference between the tensorflow and numpy

### ISSUE THREAD

**Issue by Junsong-Wang** [ppwwyyxx/tensorpack#49]
I wrote my own conv and fc layer based on numpy for data comparison of my FPGA implementation. However, I found that results from the tensorflow in your framwork and my nunpy implementation has tiny difference. For example, for the fc layer, I give the same input, while the two outputs keep a ratio of 1.0165, I don't know where does it come from? I guess it maybe from the tf.reduce_mean, which may has tiny difference from numpy's  mean function. So how I can get the E value of each layer in the inference stage, so that I can check it, or what other reason do you think can also induce this difference?, thanks.

**Comment by ppwwyyxx** [comment]
Possible reasons:
+ Models are not mathematically the same. For example, in certain cases tensorflow has a different padding mechanism from other frameworks.
+ floating point arithmetic. e.g. (a+b)/c != a/c + b/c; a * b + c with FMA instruction can have different outputs compared to simply doing multiplication and add.

If there is no problem with (1), then I'm not surprised to see a absolute difference of  ~1e-7 at first several layers and ~1e-5 at some last layers. If the error is very large you can check it layer by layer to see if some layer is mathematically different.

**Comment by Junsong-Wang** [comment]
Thanks for your comments....current I only check the FC layer, so there is no issue of padding.... I only dump the last pooling layer in the tensorflow and feed it to first FC layer in my numpy implementation. The most interesting thing is all the output keep a constant ratio.... So I doubt if it is induced by the factor E (tf.reduce_mean(tf.abs(x))), do you know how can I get this E value, so that I can double check, thanks.

**Comment by ppwwyyxx** [comment]
If you are using `output_var_names` to specify which output you want to use, you can simply `print E` to find out the name of the tensor `E` so you can run inference on this tensor.

**Comment by Junsong-Wang** [comment]
Thanks for your reply, I have confirmed that this tiny different was induced by the tf.reduce_mean, I write a simple to verify this...
```
import numpy as np
import tensorflow as tf

weight_file = 'alexnet-126.npy'
param_dict = np.load(weight_file).item()
data = param_dict['fc0/W']

print 'Numpy results:{}'.format(np.mean(np.abs(data)))

vec = tf.placeholder(tf.float32, data.shape)
avg = tf.reduce_mean(tf.abs(vec))

avgs = []
with tf.Session() as sess:
    tf_mean = sess.run(avg, feed_dict={vec: data})
    print 'Tensorflow results:{}'.format(tf_mean)
```
If run this scripts. the output is
Numpy results:0.0377448014915
Tensorflow results:0.0374231785536
It seems that the difference is much larger than I expected...such as 1e-7, what is your option?

**Comment by Junsong-Wang** [comment]
I changed the tf.float32 to tf.float64, the output seems as close as I expected....However, if I change the numpy from np.float to np.float64, it seems no change... So it seems tf.float32 has much lower precision

**Comment by ppwwyyxx** [comment]
I believe np.float is np.float64. And when you use different floating point precision this behavior is not a big surprise.

**Comment by Junsong-Wang** [comment]
It seems the difference has relation to the size of the matrix, the difference increases with the matrix size becomes large. Since the fc0 has tens millions of  weights, the difference becomes a little large... It may relate to the detail of the implementation. Do you know some detail between the tf.reduce_mean and np.mean, thanks.

**Comment by ppwwyyxx** [comment]
With floating point a+b+c != a+c+b, especially for large numbers. When there are millions of numbers it might be the case. But I also think the difference here is quite large. 
I'm closing this and you can track the relevant issue in tensorflow: https://github.com/tensorflow/tensorflow/issues/5527
